src/main/java/com/google/step/snippet/servlets/SearchServlet.java

```diff
@@ -96,14 +105,41 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
       return;
     }
     String query = encodeValue(param);
-    List<Card> allCards = new ArrayList<>();
-    for (Client client : clients) {
-      String link = getLink(client.getCseId(), query);
-      Card card = client.search(link);
-      if (card != null) {
-        allCards.add(card);
-      }
+    List<Card> allCards;
+    ExecutorService executor = Executors.newFixedThreadPool(THREAD_COUNT);
```

Comment:
So I probably should have thought about this earlier, but part of the overhead of threads is creating new threads, and we're doing that on every single request, which is probably part of why you're unlikely to see 3x speedups. My new recommendation (sorry) is to use Executors.newCachedThreadPool(), and make it a private instance variable. This way new requests reuse already existing threads whenever possible, making those requests complete faster, and we don't end up spinning up and tearing 3 threads on every request.

Reply:
For reference: https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Executors.html#newCachedThreadPool()